### PR TITLE
features for accomplices

### PR DESCRIPTION
**Pull request by patte**
### TODOs
- [x] self claimable roles
- [x] new newsletter: accomplice
- [x] hide new newsletter to users with role `accomplice`

### deploy
- [x] set ENV `MAILCHIMP_INTEREST_NEWSLETTER_ACCOMPLICE`
- [x] set ENV for republik-api `SELF_CLAIMABLE_ROLES=accomplice`

### CODE REVIEW

File: packages/auth/graphql/resolvers/_mutations/addUserToRole.js [hunk 1]
```diff
@@ -4,20 +4,28 @@ const transformUser = require('../../../lib/transformUser')
...
   if (!user) {
-    logger.error('user not found', { req: req._log() })
     throw new Error(t('api/users/404'))
   }
 
-  const returnedUser = userHasRole(user, role) ? user : (await addUserToRole(userId, role, pgdb))
+  const returnedUser = userHasRole(user, role)
+    ? user
+    : (await addUserToRole(userId, role, pgdb))
+
```
Comment (patrickvenetz) [inline packages/auth/graphql/resolvers/_mutations/addUserToRole.js]: If we made it this far, a `user` object was fetch from database, originating either from `userId` argument or `me` object. This, however, won't bother us here and we just stick to `user.id`.

```suggestion
  if (userHasRole(user, role) {
    return user
  }
  
  return addUserToRole(user.id, role, pgdb)
```

File: packages/auth/graphql/resolvers/_mutations/removeUserFromRole.js [hunk 2]
```diff
@@ -4,23 +4,28 @@ const transformUser = require('../../../lib/transformUser')
...
   if (!user) {
-    logger.error('user not found', { req: req._log() })
     throw new Error(t('api/users/404'))
   }
 
-  const returnedUser = !userHasRole(user, role) ? user : (await removeUserFromRole(userId, role, pgdb))
+  const returnedUser = !userHasRole(user, role)
+    ? user
+    : (await removeUserFromRole(userId, role, pgdb))
+
```
Comment (patrickvenetz) [inline packages/auth/graphql/resolvers/_mutations/removeUserFromRole.js]:
```suggestion
  if (!userHasRole(user, role)) {
    return user
  }
  
  return removeUserFromRole(user.id, role, pgdb)
```

File: packages/auth/lib/Roles.js [hunk 3]
```diff
@@ -96,6 +96,20 @@ const userIsMeOrProfileVisible = (user, me) => (
...
+  SELF_CLAIMABLE_ROLES
+} = process.env
+
+const isRoleClaimableByMe = (role, me) => (
+  me &&
+  role &&
+  SELF_CLAIMABLE_ROLES &&
+  SELF_CLAIMABLE_ROLES.length &&
+  SELF_CLAIMABLE_ROLES
+    .split(',')
```
Comment (patrickvenetz) [inline packages/auth/lib/Roles.js]: As in a previous request, it is likely less performant to split `SELF_CLAIMABLE_ROLES` on each fn call.

File: packages/mail/lib/getNewsletterSettings.js [hunk 4]
```diff
@@ -31,13 +34,16 @@ module.exports = async ({
 
   const status = member.status
   const subscriptions = []
-  supportedInterestIds.forEach(interestId => {
-    // only return already configured interests / subscriptions
-    if (interestId in member.interests) {
+  supportedInterestConfigs.forEach(({ interestId, visibleToRoles }) => {
+    // only return visible interests
+    const subscribed = status === MailchimpInterface.MemberStatus.Subscribed
+      ? (!!member.interests[interestId] || false)
```
Comment (patrickvenetz) [inline packages/mail/lib/getNewsletterSettings.js]: `!!` seems to make `|| false` obsolete here:

```suggestion
      ? !!member.interests[interestId]
```

File: packages/auth/graphql/resolvers/_mutations/addUserToRole.js [hunk 5]
```diff
@@ -4,20 +4,28 @@ const transformUser = require('../../../lib/transformUser')
...
+  const {
+    userId = me && me.id,
+    role
+  } = args
 
+  if (!(me && me.id === userId && isRoleClaimableByMe(role, me))) {
+    ensureUserHasRole(me, 'admin')
+  }
+
+  const user = await pgdb.public.users.findOne({ id: userId })
```
Comment (patrickvenetz) [inline packages/auth/graphql/resolvers/_mutations/addUserToRole.js]: Only fetch user from database if `me` is not user we ought to update. (And at this point we're pretty sure `me` is a user object already.)

```suggestion
  const user = me.id !== userId
    ? await pgdb.public.users.findOne({ id: userId })
    : me
```

File: packages/auth/graphql/resolvers/_mutations/removeUserFromRole.js [hunk 6]
```diff
@@ -4,23 +4,28 @@ const transformUser = require('../../../lib/transformUser')
...
-  ensureUserHasRole(req.user, 'admin')
-
-  const { userId, role } = args
-
-  const user = await pgdb.public.users.findOne({id: userId})
+  if (!(me && me.id === userId && isRoleClaimableByMe(role, me))) {
+    ensureUserHasRole(me, 'admin')
+  }
 
+  const user = await pgdb.public.users.findOne({ id: userId })
```
Comment (patrickvenetz) [inline packages/auth/graphql/resolvers/_mutations/removeUserFromRole.js]:
```suggestion
  const user = me.id !== userId
    ? await pgdb.public.users.findOne({ id: userId })
    : me
```

File: packages/mail/lib/getNewsletterSettings.js [hunk 7]
```diff
@@ -31,13 +34,16 @@ module.exports = async ({
...
+      : false
+    if (subscribed || !visibleToRoles || !visibleToRoles.length || userIsInRoles(user, visibleToRoles)) {
```
Comment (patrickvenetz) [inline packages/mail/lib/getNewsletterSettings.js]: `visibleToRoles` can be undefined. `userIsInRoles` will default undefined roles argument to an empty array (`userIsInRoles = (user, roles = []) =>...`).

Would argue to keep this concern to `userIsInRoles` fn.



```suggestion
    if (subscribed || userIsInRoles(user, visibleToRoles)) {
```

File: servers/republik/graphql/resolvers/_mutations/requestNewsletterSubscription.js [hunk 8]
```diff
@@ -16,6 +16,10 @@ module.exports = async (_, args, context) => {
     t
   } = context
 
+  if (['COVID19'].includes(name)) {
```
Comment (tpreusse) [inline servers/republik/graphql/resolvers/_mutations/requestNewsletterSubscription.js]: shouldn't it be the opposite?

Comment (patte) [reply servers/republik/graphql/resolvers/_mutations/requestNewsletterSubscription.js]: yes, thanks! I didn't check the suggestion of @patrickvenetz well enough: https://github.com/orbiting/backends/pull/400#discussion_r398692374

Comment (patte) [inline servers/republik/graphql/resolvers/_mutations/requestNewsletterSubscription.js]:
```suggestion
  if (!['COVID19'].includes(name)) {
```

File: packages/auth/lib/Roles.js (hunk 3 above)
Comment (patte) [reply packages/auth/lib/Roles.js]: I count the easier readability is much higher compared to the very small performance impact.

**Pull request merged**